pkg/reconciler/topology/partitionset_reconcile.go

```diff
@@ -0,0 +1,278 @@
+					conditionsv1alpha1.ConditionSeverityError,
+					"old partition could not get deleted",
+				)
+				return err
+			}
+			continue
+		}
+
+		// MatchLabels need to be the same
+		partitionKey := ""
+		oldMatchLabels := map[string]string{}
+		if oldPartition.Spec.Selector != nil {
+			oldMatchLabels = oldPartition.Spec.Selector.MatchLabels
+		}
+
+		// Sorting the keys of the old partition for consistent comparison
+		sortedOldKeys := make([]string, len(oldMatchLabels))
+
+		i := 0
+		for k := range oldMatchLabels {
+			sortedOldKeys[i] = k
+			i++
+		}
+		sort.Strings(sortedOldKeys)
+		for _, key := range sortedOldKeys {
+			partitionKey = partitionKey + "+" + key + "=" + oldMatchLabels[key]
+		}
+		if _, ok := matchLabelsMap[partitionKey]; ok {
+			existingMatches[partitionKey] = placeHolder
+		} else {
+			if err := c.deletePartition(ctx, logicalcluster.From(oldPartition).Path(), oldPartition.Name); err != nil {
+				conditions.MarkFalse(
+					partitionSet,
+					topologyv1alpha1.PartitionsReady,
+					topologyv1alpha1.ErrorGeneratingPartitionsReason,
+					conditionsv1alpha1.ConditionSeverityError,
+					"old partition could not get deleted",
+				)
+				return err
+			}
+		}
+	}
+	// Create partitions when no existing partition for the set has the same selector.
+	for key, matchLabels := range matchLabelsMap {
+		if _, ok := existingMatches[key]; !ok {
+			partition := generatePartition(partitionSet.Name, newMatchExpressions, matchLabels, partitionSet.Spec.Dimensions)
+			partition.OwnerReferences = []metav1.OwnerReference{
+				*metav1.NewControllerRef(partitionSet, topologyv1alpha1.SchemeGroupVersion.WithKind("PartitionSet")),
+			}
+			_, err = c.createPartition(ctx, logicalcluster.From(partitionSet).Path(), partition)
+			if err != nil && !apierrors.IsAlreadyExists(err) {
+				conditions.MarkFalse(
+					partitionSet,
+					topologyv1alpha1.PartitionsReady,
+					topologyv1alpha1.ErrorGeneratingPartitionsReason,
+					conditionsv1alpha1.ConditionSeverityError,
+					"partition could not get created",
+				)
+				return err
+			}
+		}
+	}
+	conditions.MarkTrue(partitionSet, topologyv1alpha1.PartitionsReady)
+	return nil
+}
+
+// partition populates shard label selectors according to dimensions.
+// It only keeps selectors that have at least one match.
+func partition(shards []*corev1alpha1.Shard, dimensions []string, shardSelectorLabels map[string]string) (matchLabelsMap map[string]map[string]string) {
+	matchLabelsMap = make(map[string]map[string]string)
+	labels := make([]string, len(dimensions), len(dimensions)+len(shardSelectorLabels))
+	copy(labels, dimensions)
+	for label := range shardSelectorLabels {
+		labels = append(labels, label)
+	}
+	// Sorting for consistent comparison.
+	sort.Strings(labels)
+	for _, shard := range shards {
+		key := ""
+		selector := make(map[string]string)
+		for _, label := range labels {
+			labelValue, ok := shard.Labels[label]
+			if !ok {
+				break
```

Comment:
what does it mean?
what if we didn't find let's say the second label out of four? (`len(key) > 0` will apply)

Reply:
true. Adding a boolean to check the match